[ASDataController] Refactor ASDataController

nguyenhuy opened the pull request:
## Author: Thanh Huy Nguyen

This PR has 2 main goals:
- Simplify `ASDataController` and overhaul the way it processes change sets.
  - `ASDataController` now processes a change set in 3 steps:
    - Synchronizes its internal states with data source on main thread.
    - Switches to a background queue to layout new nodes.
    - Switches back to main to forward the change set to UIKit.
  - Each step consumes the **whole** change set before switching to another thread for the next step. The data shared between these steps are immutable. This helps to guarantee thread safety and data consistency.
  - Each step may interpret a change set differently. The first step decomposes reloads into deletes and inserts, while the last step forwards the change set as-is. As a result, the implementation of `-reloadData` is much simpler. We finally can destroy everything at once and repopulate the whole data set. And flashes should be gone now (#2536). Furthermore, it's much easier to support native moves (#698, #2868) and reloads (#469).
- Merge `ASCollectionDataController` into `ASDataController`. 
  - `ASCollectionDataController` subclasses `ASDataController` to support supplementary views. In order to achieve that, it requires many hooks during the update process. Those hooks add complexity to `ASDataController`. Moreover, `ASCollectionDataController` doesn't handle supplementary node contexts correctly. It doesn't persistently store those contexts which is important for upcoming features (i.e node virtualization and measurement range).
  - Even after #1629 and #1773, repopulating supplementary nodes still doesn't work perfectly. After merging the 2 classes, it's much easier to rewrite this part.

Remaining tasks:
- [x] Fix failing unit tests
- [x] More testing on various projects
- [x] Revisit TODOs
- [x] Support reloads (#469) natively
- [ ] Support moves (#698, #2868) natively
- [ ] Revisit `-[ASDataController relayoutAllNodes]` and its callees.
- [ ] Consider changing `_nodeContexts` to be a two dimension array of dictionaries (instead of a dictionary of two dimensional arrays).

This PR is still a WIP and contains many TODOs. However, I'd appreciate any early feedbacks and help for testing. Thanks!

Ticket: https://github.com/TextureGroup/Texture/issues/186

nguyenhuy committed:
Refactor ASDataController

Check optional methods in ASDataControllerSource

nguyenhuy committed:
Reimplement reloadData

nguyenhuy committed:
Refactor new code
- No more new/inserted contexts flag
- Encapsulate code shared between reloadData and updateWithChangeSet
- Remove dataControllerWillDeleteAllData delegate method
- Hierarchy changes no longer needs to conform to NSCopying
- Reword TODOs

nguyenhuy committed:
Forgot to call completion block of reloadData :P

Adlai-Holler commented on AsyncDisplayKit/ASCellNode.h:
```diff
@@ -74,6 +74,7 @@ typedef NS_ENUM(NSUInteger, ASCellNodeVisibilityEvent) {
  */
 @property (nonatomic, assign) BOOL neverShowPlaceholders;
 
+//TODO shouldn't this be a generic kind (e.g includes ASDataControllerRowNodeKind)?
```
Since `ASDataControllerRowNodeKind` is private, we should expose that as `nil`.

nguyenhuy replied:
I guess we have a common ground that this property should be called `kind` and exposed as `nil` for row kind?

On the other hand, can `ASDataControllerRowNodeKind` possibly be exposed/public?

Adlai-Holler replied:
Yeah, I'm down for `kind` or `elementKind` 👍

Adlai-Holler commented on AsyncDisplayKit/ASTableView.mm:
```diff
@@ -657,11 +659,16 @@ - (void)beginUpdates
+  [self endUpdatesCompletion:nil];
+}
+
+- (void)endUpdatesCompletion:(void (^)(BOOL completed))completion
```
Why add this method?

nguyenhuy replied:
Right on!

nguyenhuy replied:
Right on! Removing...

Adlai-Holler commented on AsyncDisplayKit/ASTableView.mm:
```diff
@@ -1422,9 +1429,10 @@ - (void)didBeginUpdatesInRangeController:(ASRangeController *)rangeController
+  //TODO Reimplement _automaticallyAdjustsContentOffset
+//  if (_automaticallyAdjustsContentOffset) {
+//    [self beginAdjustingContentOffset];
+//  }
```
This flag has a very tumultuous history, but we should really strive to support it in every version unless we make it unavailable. If it's absolutely unbearable to implement – even in a rough way – we should be sure to put in the release notes that this flag is not functional in this version.

Adlai-Holler commented on AsyncDisplayKit/ASTableView.mm:
```diff
@@ -1454,6 +1463,33 @@ - (void)rangeController:(ASRangeController *)rangeController didEndUpdatesAnimat
+      [_rangeController updateIfNeeded];
+    }
+    // TODO Should reloadData triggers batch fetching?
+//    [self _scheduleCheckForBatchFetchingForNumberOfChanges:];
```
I believe it should. If I reload data, and I don't have e.g. 2 screenfuls of data ahead of me, I should attempt to batch fetch.

Adlai-Holler commented on AsyncDisplayKit/ASTableView.mm:
```diff
@@ -1454,6 +1463,33 @@ - (void)rangeController:(ASRangeController *)rangeController didEndUpdatesAnimat
+  
+  //TODO is _automaticallyAdjustsContentOffset needed for reloadData?
+//  if (_automaticallyAdjustsContentOffset) {
+//    [self adjustContentOffsetWithNodes:nodes atIndexPaths:indexPaths inserting:YES];
```
Hmm good question. I think it should not apply to `reloadData` because we have to assume that every single item was replaced by something completely different.

nguyenhuy replied:
Agreed!

Adlai-Holler commented on AsyncDisplayKit/Details/ASCollectionDataController.h:
```diff
@@ -1,53 +0,0 @@
-
-@end
-
-NS_ASSUME_NONNULL_END
```
WOW I'm glad to see this file gone ⚡️ 🎉 🌴

appleguy replied:
Super great!!!  Thanks @nguyenhuy!

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.h:
```diff
@@ -95,7 +112,7 @@ extern NSString * const ASCollectionInvalidUpdateException;
  Called for insertion of sections.
  */
-- (void)dataController:(ASDataController *)dataController didInsertSections:(NSArray<NSArray<ASCellNode *> *> *)sections atIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions;
+- (void)dataController:(ASDataController *)dataController didInsertSectionsAtIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions;
```
Yes! Very happy to see this argument removed. Let's phrase it `didInsertSections:(NSIndexSet *)sectionIndexes` since it's clear that we're passing an index set.

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -238,313 +241,198 @@ - (void)_batchLayoutAndInsertNodesFromContexts:(NSArray<ASIndexedNodeContext *>
+        _completedNodeContexts = loadingNodeContexts;
+        
+        // Step 3: Now that _completedNodeContexts is ready, call UIKit to update using the original change set.
+        [_delegate dataControllerDidReloadData:self];
```
Nice! Formal `reloadData`!

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -645,240 +524,194 @@ - (void)updateWithChangeSet:(_ASHierarchyChangeSet *)changeSet animated:(BOOL)an
+        _completedNodeContexts = loadingNodeContexts;
+        
+        // Step 3: Now that _completedNodeContexts is ready, call UIKit to update using the original change set.
+        [self _forwardChangeSetToDelegate:changeSet animated:animated];
```
How hard would it be to fuse `reloadData` into `updateWithChangeSet:` by adding a flag like `_ASHierarchyChangeSet.includesReloadData`? Much of the internal processing of `_ASHierarchyChangeSet` would be skipped when that flag is set, and we use it to determine e.g. how to report to the delegate.

That way there's only one implementation of this core `update` method.

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -977,91 +796,134 @@ - (ASCellNode *)nodeAtCompletedIndexPath:(NSIndexPath *)indexPath
-  ASMoveElementInTwoDimensionalArray(_externalCompletedNodes, indexPath, newIndexPath);
-  ASMoveElementInTwoDimensionalArray(_completedNodes[ASDataControllerRowNodeKind], indexPath, newIndexPath);
+  NSMutableArray<ASNodeContextTwoDimensionalArray *> *deepCopy = [NSMutableArray arrayWithCapacity:originalDict.count];
+  for (ASNodeContextTwoDimensionalArray *sections in originalDict.allValues) {
```
Rather than copying the array with `allValues`, let's use `objectEnumerator` here and below.

Adlai-Holler reviewed (commented):
This is an absolutely massive improvement in data controller! I love it, and we're gonna be able to move much much faster from here on out. Left some comments but hopefully we can land ASAP.

nguyenhuy committed:
Completion block of -[ASDataController reloadDataWithCompletion:] is nullable

nguyenhuy committed:
Data queried from ASCollectionNode and ASTableNode should be in UIKit index space
- This helps to avoid immature node allocation, especially when node virtualization is a thing
- However, this means that -reloadDataInitiallyIfNeeded in ASCollectionNode and ASTableNode must wait until all updates are finished.

nguyenhuy committed:
ASDataController shouldn't assume that allocated nodes were also laid out

nguyenhuy committed:
Revert "Data queried from ASCollectionNode and ASTableNode should be in UIKit index space"

This reverts commit 7bc977b3808a92f484b297781d0f5b30aa258e17.

nguyenhuy committed:
-nodeAtIndexPath: of ASDataController now forces node allocation, with the assumption that clients absolutely need it.
- Revisit this when node virtualization is implemented.

nguyenhuy committed:
ASDataController only grab changeSet.completionHandler when needed because it'll be niled out

nguyenhuy committed:
Fix ASTableViewTests related to reloadData

nguyenhuy committed:
Fix testThatDeletedItemsAreMarkedInvisible in ASCollectionViewTests

nguyenhuy committed:
Minor changes in ASCollectionView and ASTableView

nguyenhuy committed:
ASCollectionView and ASTableView shouldn't call [super reloadData] before their data controller does anything

nguyenhuy committed:
Address comments

nguyenhuy committed:
Fuse reloadData into -updateWithChangeSet: of ASDataController

nguyenhuy committed:
reloadData shouldn't be called as if it's inside a batch
- It can't be used in conjuntion with other updates.
- Calling it inside a batch update during the initial load can cause data inconsistency thrown by UICollectionView/UITableView

nguyenhuy committed:
Refactor ASDataControllerDelegate and ASRangeControllerDelegate
- Replace delegate methods in these protocols with -willUpdateWithChangeSet and -didUpdateWithChangeSet.
- ASRangeController, ASCollectionView and ASTableView are simplified because of this.

nguyenhuy committed:
Fix mismatch between sorting orders in ASDataController

nguyenhuy committed:
Forgot to call completion handler of reload change sets

nguyenhuy committed:
Make sure ASCollectionView is compatible with the behavior of UICollectionView's reloadData
- Since UICollectionView's reloadData doesn't requery data source but defers until the next layout pass,  we need to wait until then to update range controller and do batch fetching.
- `-[ASCollectionView waitUntilAllUpdatesAreCommited]` needs to force a layout pass to make sure
everything is ready after it returns.

nguyenhuy committed:
testSectionIndexHandling of ASTableViewTests should only check visible nodes. Other nodes will be re-measured later.

nguyenhuy committed:
ASTableView is not ready until the first layout pass finished

nguyenhuy commented:
@Adlai-Holler @appleguy This is finally ready for review! As this PR is getting huge, I'd like to merge this ASAP and work on follow up tasks. Thank you!

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -16,28 +16,40 @@
-//#define LOG(...) NSLog(__VA_ARGS__)
-#define LOG(...)
+#define LOG(...) NSLog(__VA_ARGS__)
+//#define LOG(...)
```
Turn logging back off

Adlai-Holler commented on AsyncDisplayKit/Details/ASRangeController.h:
```diff
@@ -147,64 +148,18 @@ AS_SUBCLASSING_RESTRICTED
+ * @param changeSet The change set that includes all updates
  */
-- (void)rangeController:(ASRangeController *)rangeController didDeleteItemsAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions;
+- (void)rangeController:(ASRangeController * )rangeController willUpdateWithChangeSet:(_ASHierarchyChangeSet *)changeSet;;
```
Nit: Remove the space

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.h:
```diff
@@ -76,31 +89,18 @@ extern NSString * const ASCollectionInvalidUpdateException;
+ * @param changeSet The change set that includes all updates
  */
-- (void)dataController:(ASDataController *)dataController didInsertSections:(NSArray<NSArray<ASCellNode *> *> *)sections atIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions;
+- (void)dataController:(ASDataController * )dataController willUpdateWithChangeSet:(_ASHierarchyChangeSet *)changeSet;;
```
Remove stray space

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -16,28 +16,40 @@
+// Dictionary with each entry is a pair of "kind" key and two dimensional array of node contexts
+#define ASNodeContextTwoDimensionalDictionary         NSDictionary<NSString *, ASNodeContextTwoDimensionalArray *>
+// Mutable dictionary with each entry is a pair of "kind" key and two dimensional array of node contexts
+#define ASNodeContextTwoDimensionalMutableDictionary  NSMutableDictionary<NSString *, ASNodeContextTwoDimensionalMutableArray *>
```
Nice!

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -119,30 +144,32 @@ + (NSUInteger)parallelProcessorCount
 #pragma mark - Cell Layout
 
-- (void)batchLayoutNodesFromContexts:(NSArray<ASIndexedNodeContext *> *)contexts batchCompletion:(ASDataControllerCompletionBlock)batchCompletionHandler
+- (void)batchLayoutNodesFromContexts:(NSArray<ASIndexedNodeContext *> *)contexts batchSize:(NSUInteger)batchSize batchCompletion:(ASDataControllerCompletionBlock)batchCompletionHandler
```
Nit: When adding new integer variables, let's standardize on `NSInteger` because it's simpler and the Swift import is way easier.

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -218,309 +232,137 @@ - (void)_batchLayoutAndInsertNodesFromContexts:(NSArray<ASIndexedNodeContext *>
+    // TODO: Consider using a diffing algorithm here
+    [_nodeContexts[kind] enumerateObjectsAtIndexes:sectionIndexes options:0 usingBlock:^(NSMutableArray<ASIndexedNodeContext *> * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
+      [obj removeAllObjects];
+    }];
```
Would it make more sense to do `_nodeContexts enumerateKeysAndObjectsUsingBlock:` for this removal step? That way we are sure we removed all the old supplementaries, even if that kind isn't present anymore.

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -218,309 +232,137 @@ - (void)_batchLayoutAndInsertNodesFromContexts:(NSArray<ASIndexedNodeContext *>
-- (NSArray<ASIndexedNodeContext *> *)_populateNodeContextsFromDataSourceForSections:(NSIndexSet *)sections
+- (void)_insertNodeContextsOfKind:(NSString *)kind
+                     atIndexPaths:(NSArray<NSIndexPath *> *)indexPaths
+                      environment:(id<ASTraitEnvironment>)environment
```
Nice! Love having the environment as a parameter here so we can keep track of it. In the future we should be sure that when the environment changes, we go through all affected contexts and update them. I know @rcancro was dealing with environment changes mid-update and was super frustrated.

nguyenhuy replied:
oh, that's bad, sorry @rcancro. With the new implementation, all contexts of an update are created in 1 single run loop of main thread. As a result, I don't think it's possible to have environment changes mid-update.

@Adlai-Holler Good idea! Mind creating a ticket to track it?

Adlai-Holler replied:
We have one 😛 https://github.com/facebook/AsyncDisplayKit/issues/2959

Adlai-Holler commented on AsyncDisplayKit/Details/ASDataController.mm:
```diff
@@ -546,65 +388,74 @@ - (void)invalidateDataSourceItemCounts
+  }
+  
+  ASDisplayNodeAssert(NO, @"Unknown constrained size for node of kind %@ by data source %@", kind, _dataSource);
+  return ASSizeRangeMake(CGSizeZero, CGSizeZero);
```
Nit: `ASSizeRangeZero`

Adlai-Holler commented on AsyncDisplayKit/Details/ASRangeController.h:
```diff
@@ -147,64 +148,18 @@ AS_SUBCLASSING_RESTRICTED
+ * @param changeSet The change set that includes all updates
  */
-- (void)rangeController:(ASRangeController *)rangeController didDeleteSectionsAtIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions;
+- (void)rangeController:(ASRangeController * )rangeController didUpdateWithChangeSet:(_ASHierarchyChangeSet *)changeSet;;
```
Remove the space

Adlai-Holler commented on AsyncDisplayKit/Private/_ASHierarchyChangeSet.h:
```diff
@@ -128,6 +127,9 @@ NSString *NSStringFromASHierarchyChangeType(_ASHierarchyChangeType changeType);
 
 @property (nonatomic, readonly) BOOL completed;
+/// Whether or not changes should be animated.
+@property (nonatomic, readwrite) BOOL animated;
```
Good call, putting this flag here. My reckoning here is, if any update in the stack is non-animated, the whole update should be non-animated. Would be fine as a follow-up task or discussion, or implement it here.

Adlai-Holler commented on AsyncDisplayKit/Private/_ASHierarchyChangeSet.mm:
```diff
@@ -411,6 +423,16 @@ - (void)_sortAndCoalesceChangeArrays
+    if (0 < (_originalDeleteSectionChanges.count + _originalDeleteItemChanges.count
+             +_originalInsertSectionChanges.count + _originalInsertItemChanges.count
+             + _reloadSectionChanges.count + _reloadItemChanges.count)) {
+      ASFailUpdateValidation(@"Attempt to reload data in conjuntion with other updates.");
```
I think we can probably remove this and be lenient on them. Right?

Adlai-Holler commented on AsyncDisplayKit/Private/_ASHierarchyChangeSet.mm:
```diff
@@ -536,6 +558,7 @@ - (NSString *)debugDescription
 - (NSMutableArray<NSDictionary *> *)propertiesForDescription
 {
   NSMutableArray<NSDictionary *> *result = [NSMutableArray array];
+  [result addObject:@{ @"_includesReloadData" : @(_includesReloadData) }];
```
Nit: string should just be `"includesReloadData"`

Adlai-Holler reviewed (approved):
Absolutely fantastic! I am so so happy to have these changes in. Mostly non-blocking feedback. We can build quite a bit from here! ❤️ @nguyenhuy

nguyenhuy committed:
Address comments

nguyenhuy committed:
Bug fixes

Adlai-Holler commented:
Here, we, go!
